addons/product/models/product_template.py

```diff
@@ -448,34 +447,49 @@ def create_variant_ids(self):
+            if not any(create_variant == 'dynamic' for create_variant in tmpl_id.mapped('attribute_line_ids.attribute_id.create_variant')):
+                # Iterator containing all possible attribute values combination
+                # The iterator is used to avoid MemoryError in case of a huge number of combination.
+                all_variants = itertools.product(*(
+                    line.value_ids.ids for line in tmpl_id.attribute_line_ids if line.value_ids[:1].attribute_id.create_variant != 'no_variant'
+                ))
+                # Set containing existing attribute values combination
+                existing_variants = {
+                    frozenset(variant.attribute_value_ids.filtered(lambda r: r.attribute_id.create_variant != 'no_variant').ids)
+                    for variant in tmpl_id.product_variant_ids
+                }
+                # For each possible variant, create if it doesn't exist yet.
+                for value_ids in all_variants:
+                    value_ids = set(value_ids)
+                    if value_ids not in existing_variants:
+                        variants_to_create.append({
+                            'product_tmpl_id': tmpl_id.id,
+                            'attribute_value_ids': [(6, 0, list(value_ids))],
+                        })
+                        if len(variants_to_create) > 1000:
+                            raise UserError(_("""
+                            The number of variants to generate is too high.
+                            You should either not generate variants for each combination or generate them on demand from the sales order.
+                            To do so, open the form view of attributes and change the mode of *Create Variants*."""))
+
+            # Check existing variants if any needs to be activated or unlinked.
+            # - if the product is not active and has valid attributes and attribute values, it
+            #   should be activated
+            # - if the product does not have valid attributes or attribute values, it should be
+            #   deleted
+            valid_value_ids = tmpl_id.mapped('attribute_line_ids.value_ids').filtered(
+                lambda v: v.attribute_id.create_variant != 'no_variant'
+            )
+            valid_attribute_ids = valid_value_ids.mapped('attribute_id')
             for product_id in tmpl_id.product_variant_ids:
-                if not product_id.active and product_id.attribute_value_ids.filtered(lambda r: r.attribute_id.create_variant != 'no_variant') in variant_matrix:
+                if not product_id.active and product_id._has_valid_attributes(valid_attribute_ids, valid_value_ids):
                     variants_to_activate.append(product_id)
-                elif product_id.attribute_value_ids.filtered(lambda r: r.attribute_id.create_variant != 'no_variant') not in variant_matrix:
+                elif not product_id._has_valid_attributes(valid_attribute_ids, valid_value_ids):
```

Comment:
I'd expect the "inactive variant" bit to be a pretty rare hit, so maybe test once on _has_valid_attributes and check for active within that? e.g.
```python
if product_id._has_valid_attributes(…):
    if not product_id.active:
        variants_to_activate.append(product_id)
else:
    variants_to_unlink.append(product_id)
```

Reply:
indeed, I forgot to change that